Automattic/wp-calypso
# Checkout: remove dependency on lib/cart-values product creators

#### Changes proposed in this Pull Request

This is a follow up to https://github.com/Automattic/wp-calypso/pull/46532 which moved the shopping cart manager to the `@automattic/shopping-cart` package.

In this PR we remove the dependency on the product creation functions from `lib/cart-values` from new checkout, which should hopefully allow us to eventually remove many parts of that directory and rely instead on the `@automattic/shopping-cart` package. 

Previously, checkout would use functions like `planItem()` and `domainItem()` from `lib/cart-values` to format data (mostly the product slug, but also other associated data like a domain name) into a `RequestCartProduct` prior to adding those products to the cart. Most of these functions, however, are simple mappings of the form:

```js
(product_slug, product_id, meta) => ({product_slug, product_id, meta})
```

In `usePrepareProductsForCart` (the primary consumer of these functions in composite checkout) we already have the product slug and ID and knowing what kind of product we're creating is not as important except for products which require special properties.

In this PR we add a single helper function, `createRequestCartProduct()`, which makes sure that all the required fields are set for a `RequestCartProduct` and fills in the defaults. This proved enough to replace all uses of the previously mentioned functions within the `my-sites/checkout/composite-checkout` directory.

This does not change any uses of the `lib/cart-values` functions from other parts of calypso, including those that might be used by components imported into new checkout. That will have to come later.

#### Testing instructions

This modifies how products are added to the cart from the URL when checkout loads. Therefore it's worth testing all the ways that this happens and verifying that the correct products are added each time.

- Use a URL which adds a plan to the cart via its alias like `/checkout/example.com/premium`.
- Use a URL which adds a non-plan product to the cart like `/checkout/example.com/domain-mapping:example.com`.
- Use a URL which adds a renewal product to the cart like `/checkout/value_bundle/renew/1234567/example.com`.
